src/resolve/resolved-dns-transaction.c

```diff
@@ -1843,7 +1849,23 @@ int dns_transaction_go(DnsTransaction *t) {
                 if (IN_SET(r, -EMSGSIZE, -EAGAIN))
                         r = dns_transaction_emit_tcp(t);
         }
+        if (r == -ELOOP) {
+                if (t->scope->protocol != DNS_PROTOCOL_DNS)
+                        return r;
+
+                /* One of our own stub listeners */
+                log_debug_errno(r, "Detected that specified DNS server is our own extra listener, switching DNS servers.");
+
+                dns_scope_next_dns_server(t->scope);
+
+                if (dns_scope_get_dns_server(t->scope) == t->server) {
+                        log_debug_errno(r, "Still pointing to extra listener after switching DNS servers, refusing operation.");
```

Comment:
This seems to be quite a late point to figure this out. I'd instead expect that our own servers are dropped when parsing /etc/resolv.conf (and also maybe when parsing configuration and/or data received over dns). Why would we ever want to keep such servers on our list?

Reply:
so i pondered about that, and i figured it's better to keep them around but not use them than to not even keep them around. And that's because resolved is multiple things: it's both a DNS client *and* a manager for /etc/resolv.conf. And the additional listeners are probably something that should show up in /etc/resolv.conf if we are told so (if you don't want them to be used, why have them?), and given that we likely write them out it hence made sense to allow pushing them into resolved, even if we don't want to use them ourselves.

Hope that makes sense?